Conjugating Diagrams

y-richie-y opened the pull request:
Diagrammatic conjugation will enable diagrammatic transposes, which enable "bending" and "unbending" of words.

toumix commented on discopy/drawing.py:
```diff
@@ -616,11 +616,14 @@ def draw_box(backend, positions, node, **params):
-        (right + (asymmetry if box.is_dagger else 0), height),
-        (right + (0 if box.is_dagger else asymmetry), height + .5),
-        (left, height + .5)]
+    if box.is_dagger:
```
Isn't this doing the same thing with more lines of code?

y-richie-y replied:
I've changed this to have a `points_dict` with key `(dagger, z % 2)`. Also fixed an "asymmetry" bug!

toumix commented on discopy/quantum/circuit.py:
```diff
@@ -670,6 +673,11 @@ def __init__(self, left, right):
+    @property
+    def l(self):
+        return Swap(self.right, self.left)
+    r = l
```
There's a blank line missing here.

y-richie-y replied:
`pycodestyle` says there is no style violation

toumix commented on discopy/quantum/gates.py:
```diff
@@ -36,14 +39,33 @@ def array(self):
             return self.name
-        return "QuantumGate({}, n_qubits={}, array={})".format(
+        elif self.r in GATES:
+            return self.name + '.l'
```
Maybe `.conjugate()` would be clearer in this context?
Since there is no difference between `l` and `r`.

y-richie-y replied:
created `.conjugate()` method, then set `l = r = property(conjugate)`

toumix commented on discopy/quantum/gates.py:
```diff
@@ -36,14 +39,33 @@ def array(self):
+            _conjugate=self._conjugate, _dagger=dagger)
+
+    @property
+    def r(self):
```
This method should be called `conjugate`, with `l = r = conjugate` just after.

y-richie-y replied:
I'm going to make conjugate a method, so I'll just make it return `self.l`

toumix commented on discopy/quantum/gates.py:
```diff
@@ -401,6 +438,13 @@ def array(self):
 
+    @property
+    def r(self):
+        # special case: the conjugate of RY is itself, unlike RZ and RX
```
This should be done by setting `_conjugate=None`, the same way we have `_dagger=None` for self adjoint gates.

toumix commented on discopy/quantum/gates.py:
```diff
@@ -444,6 +494,12 @@ def grad(self, var, **params):
         return _outer_prod_diag(1, 1) @ s
 
+    @property
+    def r(self):
```
You're repeating this line four times, factor it with an abstract super class!

y-richie-y replied:
done!

toumix commented on discopy/quantum/gates.py:
```diff
@@ -389,6 +421,11 @@ def array(self):
         sin, cos = self.modules.sin(half_theta), self.modules.cos(half_theta)
         return Tensor.np.array([[cos, -1j * sin], [-1j * sin, cos]])
 
+    @property
```
You're repeating this more than 5 times, factor it out!

toumix commented on discopy/quantum/gates.py:
```diff
@@ -501,6 +563,13 @@ def grad(self, var, **params):
+        algebraic_conj = type(self)(-self.phase)
+        return Swap(qubit, qubit) >> algebraic_conj >> Swap(qubit, qubit)
+
+    l = r
```
This `l = r` should only need to appear once!

toumix commented on discopy/rigid.py:
```diff
@@ -151,6 +165,28 @@ class Diagram(monoidal.Diagram):
     .. image:: ../_static/imgs/rigid/diagram-example.png
         :align: center
     """
+    def __init__(self, dom, cod, boxes, offsets, layers=None):
```
Wait... Did you just copy and paste the code from monoidal? I don't understand why you need this.

y-richie-y replied:
removed it

toumix commented on discopy/rigid.py:
```diff
@@ -249,6 +285,55 @@ def curry(diagram, n_wires=1, left=False):
+        list_of_layers = []
+        for layer in layers._boxes:
+            layer_r = layer.r
+            left, box, right = layer_r._left, layer_r._box, layer_r._right
```
This can be done with `left, box, right = layer`.

toumix commented on discopy/rigid.py:
```diff
@@ -249,6 +285,55 @@ def curry(diagram, n_wires=1, left=False):
+        return self.upgrade(Diagram(*inputs, layers=layers_l))
+
+    @property
+    def r(self):
```
Is there a way to factor the common code for `l` and `r`? This is a lot of code duplication.

y-richie-y replied:
refactored using `._adjoint(use_left)`.

toumix commented on discopy/rigid.py:
```diff
@@ -313,6 +405,39 @@ class Box(monoidal.Box, Diagram):
         Diagram.__init__(self, dom, cod, [self], [0], layers=self.layers)
+        self._z = params.get("_z", 0)
+
+    def __eq__(self, other):
```
Is this all copied from monoidal?
We should get rid of the `Id` class altogether at some point. It would make our lives a lot easier to make it just a method of the different `Diagram` classes.

toumix commented on discopy/rigid.py:
```diff
@@ -434,8 +575,28 @@ def adjoint(obj):
+            if not hasattr(diagram, "z") or not diagram.z:
+                return super().__call__(diagram)
+            z = diagram.z
+            for _ in range(-z):
```
You can put the two loops together using `abs(z)` and an `if` statement inside.

toumix commented on discopy/rigid.py:
```diff
@@ -434,8 +575,28 @@ def adjoint(obj):
         if isinstance(diagram, monoidal.Diagram):
-            return super().__call__(diagram)
+            scan, result = diagram.dom, self.ar_factory.id(self(diagram.dom))
+            for box, off in zip(diagram.boxes, diagram.offsets):
```
This looks like it's a copy and paste from monoidal, use `super().__call__` instead?

y-richie-y commented:
@toumix I've addressed all the issues and implemented drawing and conjugation for tensor module as well. If no other changes are needed I will merge this into main.
